### PR TITLE
Updated fileUploaderservice cloneFile method to support clone file using URL.

### PR DESCRIPTION
**Pull request by laxmisalunkhe**

### CODE REVIEW

File: grails-app/services/com/lucastex/grails/fileuploader/FileUploaderService.groovy [hunk 1]
```diff
@@ -341,32 +341,45 @@ class FileUploaderService {
...
         def destFile = new File(tempFile)
-        def sourceFile = new File(ufileInstance.path)
         if (!destFile.exists()) {
             destFile.createNewFile()
         }
 
-        FileChannel source = null
-        FileChannel destination = null
+        String sourceFilePath = ufileInstance.path
+        if (urlValidator.isValid(sourceFilePath)) {
```
Comment (SAgrawal14) [inline grails-app/services/com/lucastex/grails/fileuploader/FileUploaderService.groovy]: Actually you have to check for the `type` field in `UFile` domain to check if it is a cloud based ufile. Url validator is a good option but that needs to be in secondary check.

File: grails-app/services/com/lucastex/grails/fileuploader/FileUploaderService.groovy [hunk 2]
```diff
@@ -341,32 +341,45 @@ class FileUploaderService {
...
+            try {
+                source = new FileInputStream(sourceFile).getChannel()
+                destination = new FileOutputStream(destFile).getChannel()
+                destination.transferFrom(source, 0, source.size())
+            } finally {
+                source?.close()
+                destination?.close()
             }
         }
+        if (destFile.exists()) {
```
Comment (SAgrawal14) [inline grails-app/services/com/lucastex/grails/fileuploader/FileUploaderService.groovy]: Remove this `if` condition, this must be handled by `saveFile` merthod.

File: grails-app/services/com/lucastex/grails/fileuploader/FileUploaderService.groovy (hunk 1 above)
Comment (laxmisalunkhe) [inline grails-app/services/com/lucastex/grails/fileuploader/FileUploaderService.groovy]: Added aditional check for UFile type filed also.

File: grails-app/services/com/lucastex/grails/fileuploader/FileUploaderService.groovy [hunk 3]
```diff
@@ -386,8 +396,10 @@ class FileUploaderService {
         List<BlobDetail> blobDetailList = []
 
         ufileInstanceList.each {
-            String newFileName = "${it.fileGroup}-${System.currentTimeMillis()}-${it.fullName}"
+            String fullName = it.fullName.trim().replaceAll(" ", "_").replaceAll("-", "_")
+            String newFileName = "${it.fileGroup}-${System.currentTimeMillis()}-${fullName}"
             blobDetailList << new BlobDetail(newFileName, new File(it.path), it.id)
+            Thread.sleep(2)
```
Comment (visheshd) [inline grails-app/services/com/lucastex/grails/fileuploader/FileUploaderService.groovy]: Is this needed?

Comment (laxmisalunkhe) [inline grails-app/services/com/lucastex/grails/fileuploader/FileUploaderService.groovy]: Removed, not needed now.

**Comment by laxmisalunkhe** [comment]
Created pull request to latest release #27

**Pull request closed without merging**